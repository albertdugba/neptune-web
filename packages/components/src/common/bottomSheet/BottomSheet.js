import React, { forwardRef } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

import Dimmer from '../../dimmer';
import SlidingPanel from '../../slidingPanel';
import './BottomSheet.css';
import { Position } from '../propsValues/position';

const BottomSheet = forwardRef(
  ({ children, open, className, onClose }, ref) =>
    open && (
      <Dimmer open={open} onClose={onClose}>
        <SlidingPanel
          open={open}
          position={Position.BOTTOM}
          className={classNames('np-bottom-sheet', className)}
          ref={ref}
        >
          {children}
        </SlidingPanel>
      </Dimmer>
    ),
);

BottomSheet.propTypes = {
  children: PropTypes.node.isRequired,
  className: PropTypes.string,
  onClose: PropTypes.func.isRequired,
  open: PropTypes.bool,
};

BottomSheet.defaultProps = {
  className: undefined,
  open: false,
};

export default BottomSheet;
