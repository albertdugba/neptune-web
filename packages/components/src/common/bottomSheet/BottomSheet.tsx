import { useRef, useState, ReactElement, PropsWithChildren, CSSProperties } from 'react';

import classNames from 'classnames';

import Dimmer from '../../dimmer';
import SlidingPanel from '../../slidingPanel';
import CloseButton from '../closeButton';

import { isServerSideRendering } from '../domHelpers';
import { useConditionalListener } from '../hooks';

import { Position } from '../propsValues/position';
import { CommonProps } from '../commonProps';

const INITIAL_Y_POSITION = 0;

const CONTENT_SCROLL_THRESHOLD = 1;

const MOVE_OFFSET_THRESHOLD = 50;

type Props = PropsWithChildren<{ onClose: () => void; open: boolean } & CommonProps>;

/**
 * Neptune: https://transferwise.github.io/neptune/components/bottom-sheet/
 *
 * Neptune Web: https://transferwise.github.io/neptune-web/components/overlays/BottomSheet
 */
function BottomSheet(props: Props): ReactElement {
  const bottomSheetRef = useRef<HTMLDivElement>(null);
  const topBarRef = useRef<HTMLDivElement>(null);
  const contentRef = useRef<HTMLDivElement>(null);

  const [pressed, setPressed] = useState<boolean>(false);

  /**
   * Used to track `requestAnimationFrame` requests
   *
   * @see https://developer.mozilla.org/en-US/docs/Web/API/window/requestAnimationFrame#return_value
   */
  const animationId = useRef<number>(0);

  /**
   * Difference between initial coordinate ({@link initialYCoordinate})
   * and new position (when user moves component), it's get calculated on `onTouchMove` and `onMouseMove` events
   *
   * @see {@link calculateOffsetAfterMove}
   */
  const moveOffset = useRef<number>(0);
  const initialYCoordinate = useRef<number>(0);

  // apply shadow to the bottom of top-bar when scroll over content
  useConditionalListener({
    attachListener: props.open && !isServerSideRendering(),
    callback: () => {
      if (topBarRef.current !== null) {
        const { classList } = topBarRef.current;
        if (!isContentScrollPositionAtTop()) {
          classList.add('np-bottom-sheet--top-bar--shadow');
        } else {
          classList.remove('np-bottom-sheet--top-bar--shadow');
        }
      }
    },
    eventType: 'scroll',
    parent: isServerSideRendering() ? undefined : document,
  });

  function move(newHeight: number): void {
    if (bottomSheetRef.current !== null) {
      bottomSheetRef.current.style.transform = `translateY(${newHeight}px)`;
    }
  }

  function close(): void {
    setPressed(false);
    moveOffset.current = INITIAL_Y_POSITION;
    if (bottomSheetRef.current !== null) {
      bottomSheetRef.current.style.removeProperty('transform');
    }
    if (props.onClose) {
      props.onClose();
    }
  }

  const onSwipeStart = (isMobile: boolean) => (event: TouchEvent & MouseEvent): void => {
    initialYCoordinate.current = (isMobile ? event.touches[0] : event).clientY;
    setPressed(true);
  };

  const onSwipeMove = (isMobile: boolean) => (event: TouchEvent & MouseEvent): void => {
    if (pressed) {
      const { clientY } = isMobile ? event.touches[0] : event;

      const offset = calculateOffsetAfterMove(clientY);
      // check whether move is to the bottom only and content scroll position is at the top
      if (offset > INITIAL_Y_POSITION && isContentScrollPositionAtTop()) {
        moveOffset.current = offset;
        animationId.current = requestAnimationFrame(() => {
          if (animationId.current !== undefined && bottomSheetRef.current !== null) {
            move(offset);
          }
        });
      }
    }
  };

  function onSwipeEnd(): void {
    // stop moving component
    cancelAnimationFrame(animationId.current);
    setPressed(false);
    // check whether move down is strong enough
    // and content scroll position is at the top to close the component
    if (moveOffset.current > MOVE_OFFSET_THRESHOLD && isContentScrollPositionAtTop()) {
      close();
    }
    // otherwise move component back to default (initial) position
    else {
      move(INITIAL_Y_POSITION);
    }
    moveOffset.current = INITIAL_Y_POSITION;
  }

  function isContentScrollPositionAtTop(): boolean {
    return (
      contentRef?.current?.scrollTop !== undefined &&
      contentRef.current.scrollTop <= CONTENT_SCROLL_THRESHOLD
    );
  }

  /**
   * Calculates how hard user moves component,
   * result value used to determine whether to hide component or re-position to default state
   */
  function calculateOffsetAfterMove(afterMoveYCoordinate: number): number {
    return afterMoveYCoordinate - initialYCoordinate.current;
  }

  /**
   * Set `max-height` for content part (in order to keep it scrollable for content overflow cases) of the component
   * and ensures space for safe zone (32px) at the top.
   */
  function setContentMaxHeight(): CSSProperties {
    const safeZoneHeight: string = '64px';
    const topbarHeight: string = '32px';
    const windowHight: number = isServerSideRendering() ? 0 : window.innerHeight;
    /**
     * Calculate _real_ height of the screen (taking into account parts of browser interface).
     *
     * See https://css-tricks.com/the-trick-to-viewport-units-on-mobile for more details.
     */
    const screenHeight: string = `${windowHight * 0.01 * 100}px`;
    return {
      maxHeight: `calc(${screenHeight} - ${safeZoneHeight} - ${topbarHeight})`,
    };
  }
  return (
    <Dimmer open={props.open} onClose={close} fadeContentOnEnter fadeContentOnExit>
      {/* @ts-expect-error remove when SlidingPanel will be written on TS */}
      <SlidingPanel
        open={props.open}
        position={Position.BOTTOM}
        ref={bottomSheetRef}
        className={classNames('np-bottom-sheet', props.className)}
        onTouchStart={onSwipeStart(true)}
        onTouchMove={onSwipeMove(true)}
        onTouchEnd={onSwipeEnd}
        onMouseDown={onSwipeStart(false)}
        onMouseMove={onSwipeMove(false)}
        onMouseUp={onSwipeEnd}
        role="dialog"
      >
        <div className="np-bottom-sheet--top-bar" ref={topBarRef}>
          <div className="np-bottom-sheet--handler" />
          {/* @ts-expect-error remove when CloseButton will be written on TS */}
          <CloseButton onClick={close} size={16} className="sr-only np-bottom-sheet--close-btn" />
        </div>
        <div style={setContentMaxHeight()} className="np-bottom-sheet--content" ref={contentRef}>
          {props.children}
        </div>
      </SlidingPanel>
    </Dimmer>
  );
}

export default BottomSheet;
