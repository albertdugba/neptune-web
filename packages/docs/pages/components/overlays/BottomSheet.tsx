import { BottomSheet, Button, NavigationOption, Size } from '@transferwise/components';
import { Card, PinCode, FastFlag, Menu } from '@transferwise/icons';
import { ReactElement, useState } from 'react';

import Heading from '../../../components/Heading';
// @ts-expect-error liveEditorCode
import code from '../../../liveEditorCode/bottomSheet.code';
import { LiveEditorBlock, GeneratePropsTable } from '../../../utils';
import { Meta } from '../../../utils/sections';

const PageContent = (): ReactElement => {
  const [openVarientOne, setOpenVarientOne] = useState(false);
  const [openVarientTwo, setOpenVarientTwo] = useState(false);
  return (
    <div>
      <p>
        A bottom sheet is a container you can use to display supporting content or a short,
        supplementary task on <b>mobile experiences</b>. See{' '}
        {/* TODO replace with Link component from `@transferwise/components` */}
        <a
          href="https://transferwise.github.io/neptune/components/bottom-sheet/"
          target="_blank"
          rel="noreferrer"
        >
          usage guidelines
        </a>{' '}
        for more info.
      </p>
      <Heading as="h2">Variants</Heading>
      <ul>
        <li>
          Hugs Content{' '}
          <Button
            size={Size.SMALL}
            onClick={() => {
              setOpenVarientOne(true);
            }}
          >
            Demo
          </Button>
        </li>
        <BottomSheet open={openVarientOne} onClose={() => setOpenVarientOne(false)}>
          <Heading as="h2" className="p-x-2">
            Settings
          </Heading>
          <NavigationOption
            title="Show me insights for"
            content="All activity"
            showMediaAtAllSizes
            media={<FastFlag size={24} />}
            className="p-l-2"
          />
          <NavigationOption
            showMediaAtAllSizes
            title="Show PIN"
            media={<PinCode size={24} />}
            className="p-l-2"
          />
        </BottomSheet>
        <li>
          Content overflow{' '}
          <Button
            size={Size.SMALL}
            onClick={() => {
              setOpenVarientTwo(true);
            }}
          >
            Demo
          </Button>
        </li>
        <BottomSheet open={openVarientTwo} onClose={() => setOpenVarientTwo(false)}>
          <Heading as="h2" className="p-x-2">
            Money without borders
          </Heading>
          <div className="p-x-2">
            Lorem ipsum dolor sit amet consectetur adipisicing elit. Hic modi sed possimus, minima
            dolorem qui suscipit omnis commodi accusamus soluta exercitationem iste reprehenderit
            assumenda repudiandae minus ducimus perspiciatis, pariatur consectetur incidunt repellat
            accusantium? Ipsum itaque placeat corporis ipsa, esse sed. Magnam ratione eum ullam
            consequatur, nulla tenetur repellendus quo eaque molestiae ducimus soluta sunt dolor
            dicta autem qui vitae alias. Facilis laboriosam ex obcaecati, assumenda alias rem
            voluptatum minus delectus vero, distinctio tempora est mollitia iste deleniti autem
            dignissimos doloremque ratione! Dicta voluptas doloremque, harum deserunt corrupti aut
            tempore nobis cum obcaecati adipisci eveniet esse, quis totam alias excepturi hic. Lorem
            ipsum dolor sit amet consectetur adipisicing elit. Hic modi sed possimus, minima dolorem
            qui suscipit omnis commodi accusamus soluta exercitationem iste reprehenderit assumenda
            repudiandae minus ducimus perspiciatis, pariatur consectetur incidunt repellat
            accusantium? Ipsum itaque placeat corporis ipsa, esse sed. Magnam ratione eum ullam
            consequatur, nulla tenetur repellendus quo eaque molestiae ducimus soluta sunt dolor
            dicta autem qui vitae alias. Facilis laboriosam ex obcaecati, assumenda alias rem
            voluptatum minus delectus vero, distinctio tempora est mollitia iste deleniti autem
            dignissimos doloremque ratione! Dicta voluptas doloremque, harum deserunt corrupti aut
            tempore nobis cum obcaecati adipisci eveniet esse, quis totam alias excepturi hic. Lorem
            ipsum dolor sit amet consectetur adipisicing elit. Hic modi sed possimus, minima dolorem
            qui suscipit omnis commodi accusamus soluta exercitationem iste reprehenderit assumenda
            repudiandae minus ducimus perspiciatis, pariatur consectetur incidunt repellat
            accusantium? Ipsum itaque placeat corporis ipsa, esse sed. Magnam ratione eum ullam
            consequatur, nulla tenetur repellendus quo eaque molestiae ducimus soluta sunt dolor
            dicta autem qui vitae alias. Facilis laboriosam ex obcaecati, assumenda alias rem
            voluptatum minus delectus vero, distinctio tempora est mollitia iste deleniti autem
            dignissimos doloremque ratione! Dicta voluptas doloremque, harum deserunt corrupti aut
            tempore nobis cum obcaecati adipisci eveniet esse, quis totam alias excepturi hic.
          </div>
        </BottomSheet>
      </ul>
      <LiveEditorBlock
        code={code as string}
        scope={{ BottomSheet, Button, NavigationOption, Size, PinCode, Card, Menu }}
      />
      <GeneratePropsTable componentName="BottomSheet" />
      <br />
      <br />
    </div>
  );
};

export const meta: Meta = {
  name: 'BottomSheet',
  badge: {
    expiryDate: '2021-09-30',
    type: 'new',
  },
};

export default PageContent;
