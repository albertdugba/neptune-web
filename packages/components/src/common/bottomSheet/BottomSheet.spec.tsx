import { render } from '../../test-utils';

import BottomSheet from './BottomSheet';

describe('BottomSheet', () => {
  it('renders content when open', () => {
    const { baseElement } = render(
      <BottomSheet open onClose={jest.fn()}>
        <b>Test Content</b>
      </BottomSheet>,
    );
    expect(baseElement).toMatchSnapshot();
  });

  it('does not renders when closed', () => {
    const { container } = render(
      <BottomSheet open={false} onClose={jest.fn()}>
        content
      </BottomSheet>,
    );
    expect(container).toBeEmptyDOMElement();
  });
});
