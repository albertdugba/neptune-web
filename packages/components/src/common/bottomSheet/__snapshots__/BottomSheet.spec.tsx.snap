// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BottomSheet renders content when open 1`] = `
<body
  class="no-scroll"
>
  <div />
  <div
    class="dimmer"
    role="presentation"
  >
    <span
      class="np-focus-boundary outline-none"
      tabindex="-1"
    >
      <div
        class="sliding-panel np-bottom-sheet sliding-panel--open-bottom sliding-panel-appear sliding-panel-appear-active"
        role="dialog"
      >
        <div
          class="np-bottom-sheet--top-bar"
        >
          <div
            class="np-bottom-sheet--handler"
          />
          <button
            aria-label="Close"
            class="np-close-button close btn-link text-no-decoration sr-only np-bottom-sheet--close-btn"
            type="button"
          >
            <span
              aria-hidden="true"
              class="tw-icon tw-icon-cross "
              role="presentation"
            >
              <svg
                fill="currentColor"
                focusable="false"
                height="16"
                width="16"
              >
                <path
                  d="M9.131 8l5.435-5.434-1.132-1.132L8 6.87 2.566 1.434 1.434 2.566 6.87 8l-5.435 5.434 1.132 1.132L8 9.13l5.434 5.435 1.132-1.132L9.13 8z"
                />
              </svg>
            </span>
          </button>
        </div>
        <div
          class="np-bottom-sheet--content"
          style="max-height: calc(768px - 64px - 32px);"
        >
          <b>
            Test Content
          </b>
        </div>
      </div>
    </span>
  </div>
</body>
`;
